### ISSUE TITLE
Different version of static or shared HDF5 library

### ISSUE THREAD

**Issue by safalabolo** [yannforget/conda-recipes#1]
Dear Yann,
Thanks again for your answers.
Currently it seems to be an incompatibility between the version of the library HDF5 used by my installation and file format.
If you can take a sguaro the attached log files.
Thanks for listening and the support.
All the best
[log.txt](https://github.com/yannforget/conda-recipes/files/371824/log.txt)

**Comment by yannforget** [comment]
Hi,
Yes, it's the same issue. You need the `hdf5` package from `conda-forge`.

```
conda install -c conda-forge hdf5
```

You can also try `export HDF5_DISABLE_VERSION_CHECK=1`.

If you want to avoid these issues and keep a clean environment, you can create a new conda virtual environment for ledaps/lasrc/cfmask. For example, for a conda env named `landsat`:

``` bash
# Create a new conda environment
conda create --name landsat python

# Activate the environment
source activate landsat

# Add custom conda channels if necessary
conda config --add channels conda-forge
conda config --add channels yannforget

# Install conda packages
conda install ledaps lasrc cfmask

# Do some processing

# Quit the environment
source deactivate
```

Normally, if the `conda-forge` channel is registered, conda will select the correct `hdf5` and `hdf4` packages since newer versions are only available in `conda-forge`. But if you want to be sure that the correct packages are installed:

``` bash
conda install -c conda-forge hdf5 hdf4
```

**Comment by safalabolo** [comment]
Dear Yann,
Updating the reanalysis data seems to work!

I would like to update all the ancillary data and then perform a first run.
I imagine that the command sequence for generating CDR products is that described in sub-paragraphs of the following links:
https://github.com/USGS-EROS/espa-surface-reflectance/blob/master/ledaps/ledapsSrc/doc/LEDAPS_User_Ref_SR.txt

I update you on next steps.
Thanks for the support, at the moment is very much appreciated.

**Comment by yannforget** [comment]
Indeed, you can follow the official doc.

There is also some python wrapper scripts : `do_ledaps.py` or `surface_reflectance.py`. These are python 2 scripts, so the default python version of your conda installation has to be 2.7 and not 3.5. If it's not the case, a workaround could be to edit the shebang of the script you want to run in order to the python2 installation of your distribution:

``` bash
# Activate the conda env if necessary
source activate <env_name>

# Find the path of the script
which do_ledaps.py

# Change the shebang
sed -i '1s:.*:#!/usr/bin/env python2:' do_ledaps.py
```

**Comment by safalabolo** [comment]
Dear Yann,

#env for me is landsat
safalabolo@safalabolo-HPE-340it:~$ source activate landsat
(landsat) safalabolo@safalabolo-HPE-340it:~$ which do_ledaps.py


(landsat) safalabolo@safalabolo-HPE-340it:~$ sed -i '1s:.*:#!/usr/bin/landsat  python2:' do_ledaps.py
sed: unable to read do_ledaps.py: File or directory does not exist
(landsat) safalabolo@safalabolo-HPE-340it:~$ 

I tried to search the files do_ledaps.py or surface_reflectance.py
but without any positive result.
Thanks!

**Comment by yannforget** [comment]
My bad, these scripts aren't integrated for now. So you have to run `convert_lpgs_to_espa` (if necessary), `lndpm`, `lndcal`, `lndcsm` and finally `lndsr` (in this order)

**Comment by safalabolo** [comment]
Dear Yann,
I wanted to process an landsat-8 image, but this is not possible?
I try with a landsat 7...
Thanks in advance for your reply.


[log_LT8.txt](https://github.com/yannforget/conda-recipes/files/373722/log_LT8.txt)

**Comment by safalabolo** [comment]
Dear Yann,
I tried to process Landsat-7 image: it seems to work! 
(I made a request of that acquisition produced by ESPA, to compare my output with the product via Web).
The only "problem" seems to be the lndcsm command because I do not find as for the other (convert_lpgs_to_espa, lndpm, lndcal ... lndsr) a txt or xml files to aim.
Thanks in advance for your reply.
All the best
[log_LT7.txt](https://github.com/yannforget/conda-recipes/files/373780/log_LT7.txt)

**Comment by yannforget** [comment]
You can only preprocess Landsat 5 & 7 scenes with Ledaps. For Landsat 8, you need LaSRC.

It seems that `lndcsm` function has been merged into `lndsr`, so in my opinion it's not an issue. However, you certainly will have to to use `convert_espa_to_gtiff` command if you prefer to work with Geotiff files.

**Comment by safalabolo** [comment]
Dear Yann,
link to auxilliary files for Landsat-8 is broken?
Thanks!

**Comment by yannforget** [comment]
Seems to be broken yes :(

https://github.com/USGS-EROS/espa-surface-reflectance/issues/48

**Comment by safalabolo** [comment]
Dear Yann,
finally the auxiliary data for landsat-8 are available!
Unfortunately in the installation folders I do not find it scritp updatelads.py for updating ancillary data.
The steps in
https://github.com/USGS-EROS/espa-surface-reflectance/tree/master/not-validated-prototype-lasrc 
is out of date?
Thanks for the answer

**Comment by yannforget** [comment]
Thanks for your input,
The conda package is updated to install the aux scripts. You can upgrade your package with: `conda install lasrc`.

The steps are not out of date, but AFAIK there has been somes changes with the related FTP services. Doesn't know if the `updatelads.py` works correctly today... I'm currently working on my own version of this script.

**Comment by safalabolo** [comment]
Dear Yann,
thanks for you update!
I have simply downloaded the python script from github, but the upgrade of your package is better action!
I'm waiting for username and password from USGS EROS  required for access to ancillary data.
Thanks for your support!
Salvatore

**Comment by safalabolo** [comment]
Dear Yann,
the module is not loaded, maybe the commands executed by me are not correct or appropriate to open a new environment such landsat1? ... but as you can delete all the previous?
Thanks in advance for yor support
[log.txt](https://github.com/yannforget/conda-recipes/files/376495/log.txt)

**Comment by safalabolo** [comment]
I forgot ...I received  username and password from USGS EROS.

**Comment by yannforget** [comment]
You need the `requests` python module. I forgot to add it as a dependency in the conda package. You can solve the issue by (1) upgrading to the updated conda package with `conda install lasrc` or (2) just install requests with `conda install requests`.

Good luck with the `updatelads.py` script ! Doesn't work for me.

**Comment by safalabolo** [comment]
Dear Yann,
maybe the problem is in the address ftp

I find in the script

     Server_url = 'ladssci.nascom.nasa.gov'
     TERRA_CMA = '/ ALLDATA / 6 / MOD09CMA /'
     TERRA_CMG = '/ ALLDATA / 6 / MOD09CMG /'
     AQUA_CMA = '/ ALLDATA / 6 / MYD09CMA /'
     AQUA_CMG = '/ ALLDATA / 6 / MYD09CMG /'

in fact it should be

     Server_url = 'ladssci.nascom.nasa.gov'
     TERRA_CMA = '/ 6 / MOD09CMA /'
     TERRA_CMG = '/ 6 / MOD09CMG /'
     AQUA_CMA = '/ 6 / MYD09CMA /'
     AQUA_CMG = '/ 6 / MYD09CMG /'

then to the latest version (6) of the MODIS products.
But maybe I'm wrong.
All the best.
Salvatore

**Comment by yannforget** [comment]
Yes,

The issue is that CMG products are not available on the ladssci FTP. But they are available here : http://e4ftl01.cr.usgs.gov. You need to register a NASA EarthData account in addition to the ladssci credentials, that's not very convenient but it's the only way I found.

One you have Terra/Aqua CMG/CMA products (4 .hdf files), you have to combine them with `combine_l8_aux_data` to generate the file that is used by `lasrc`. I have a script for the whole process (download related images and combine them for a given date). It is a little dirty for now, but if you need it urgently I can publish the current version.

**Comment by safalabolo** [comment]
Dear Yann,
Thanks for the updates.
It is as you say!
Yes, the code it yourself product to solve this problem I would be of great help!!

**Comment by yannforget** [comment]
[Here it is](https://github.com/yannforget/landsat-scripts)

You have to register an [Earth Data account](https://urs.earthdata.nasa.gov/users/new). I think there is also a supplementary step if you want to download products, like authorizing some access to your Earth Data account... Try to download a random file [here](http://e4ftl01.cr.usgs.gov/MOLT/MOD09CMG.005/) and you will certainly be prompted about that.

**Comment by safalabolo** [comment]
Dear Yann,
Thanks for your support.
I can not interpret the error
It may be the sequence of commands?
or the python version I used, the 2.7.12?
Please find the attached log file
[log_.txt](https://github.com/yannforget/conda-recipes/files/378205/log_.txt)

**Comment by yannforget** [comment]
Yes, you need Python 3 to run this script. There is a python3 shebang on the top of the script. I think you can easily install Python 3 on Ubuntu, if it's not installed by default?

Try:
`python3 <script> <options>`

**Comment by yannforget** [comment]
You need Python3 to run this script.

There is a Python3 shebang on top of the script (`#!/usr/bin/env python3`). By running the script through `python <script>`, you are overriding the shebang. If you don't specify any interpreter it will uses python3 by default (`./landsat_download_lads.py <options>`). You can also force the use of python3 by running `python3 <script>`.

Maybe you will need the python3-requests package too.

**Comment by safalabolo** [comment]
Dear Yann, 
unfortunately still an error...

export L8_AUX_DIR="/home/safalabolo/anaconda2/l8sr_auxiliary"

python3 landsat_download_laads.py --date 2016-07-21 --euser XXX --epass XXX --luser XXX --lpass XXX

Traceback (most recent call last):
  File "landsat_download_laads.py", line 85, in <module>
    args.luser, args.lpass)
  File "landsat_download_laads.py", line 55, in download_laads
    _download_cmg(date, data_dir, earthdata_username, earthdata_password)
  File "landsat_download_laads.py", line 33, in _download_cmg
    remote_fn = regex.search(dl_page).group(0)
AttributeError: 'NoneType' object has no attribute 'group'

Thanks

**Comment by yannforget** [comment]
CMG & CMA products for 07-21 are not yet available :) Lastest are 07-20. See [here](http://e4ftl01.cr.usgs.gov/MOLT/MOD09CMG.006/)

**Comment by safalabolo** [comment]
Dear Yann,
I apologize for the silence.
The atmospheric correction of landsat 7:08 works.
The only problem is the acquisition of CMG & CMA products for Landsat-8, problems that arise in the execution of the script 8landsat_download_laads.py) and the long waiting before to find them on their ftp
The solution to this problem is to download the data by hand (when present) and then combine them (irrespective of the performance of the script).

I am attaching a possible workflow for the processing of acquisitions landsat 4-5-7 and one for Landsat-8, in case you have suggestions from you.
Thanks in advance deposit for the response and support
[workflow_LT.txt](https://github.com/yannforget/conda-recipes/files/385789/workflow_LT.txt)

**Comment by safalabolo** [comment]
Dear Yann,
I apologize for the silence.
The atmospheric correction of landsat 7:08 works.
The only problem is the acquisition of CMG & CMA products for Landsat-8, problems that arise in the execution of the script 8landsat_download_laads.py) and the long waiting before to find them on their ftp
The solution to this problem is to download the data by hand (when present) and then combine them (irrespective of the performance of the script).

I am attaching a possible workflow for the processing of acquisitions landsat 4-5-7 and one for Landsat-8, in case you have suggestions from you.
Thanks in advance for your response and for support
[workflow_LT.txt](https://github.com/yannforget/conda-recipes/files/385805/workflow_LT.txt)

**Comment by yannforget** [comment]
@safalabolo ,
Yes, acquisition of auxiliary products for Landsat 8 is the main challenge to run LaSRC. You say that the script doesn't work for you, what is the output ?

Workflow seems good to me, some suggestions:
* It's not necessary to run `updatencep.py` and `updatetoms.py` if you don't use 2014+ landsat 7 data (don't know if it's your case) ;
* You must define `LEDAPS_AUX_DIR` before running ledaps ;
* You must define `L8_AUX_DIR` before running LaSRC (instead of ledaps_aux_dir in your workflow) ;
* It mights also be necessary to define `ESPA_LAND_MASS_POLYGON`

**Comment by safalabolo** [comment]
Dear Yann,
thanks for your feedback.

Yes in my workflow he was correct
export L8_AUX_DIR = "/ home / safalabolo / anaconda2 / l8sr_auxiliary"
for LaSRC.
Now I do not know what is more correct
export ESPA_LAND_MASS_POLYGON = / home / safalabolo / anaconda2 / static_data / land_no_buf.ply
or
export ESPA_LAND_MASS_POLYGON = "/ home / safalabolo / anaconda2 / static_data / land_no_buf.ply"

It is possible to define once and for all the export option in the bashrc file?
Thank!
[workflow_LT_up.txt](https://github.com/yannforget/conda-recipes/files/386579/workflow_LT_up.txt)

**Comment by yannforget** [comment]
Both are correct, with or without quotes. And yes, you can define these environment variables in your `.bashrc`
